### PR TITLE
dep: repo: handle dict as config

### PR DESCRIPTION
**Pull request by efiop**
https://discord.com/channels/485586884165107732/485596304961962003/1141530399982485574

Kudos @sjawhar 🙏

### CODE REVIEW

File: dvc/dependency/repo.py [hunk 1]
```diff
@@ -114,7 +114,9 @@ def _make_fs(
             remote_config = None
 
         conf = self.def_repo.get("config")
-        if conf:
+        if isinstance(conf, dict):
+            config = conf
```
Comment (efiop) [inline dvc/dependency/repo.py:118]:
```suggestion
            config = conf  # type: ignore[unreachable]
```

File: dvc/dependency/repo.py [hunk 2]
```diff
@@ -114,7 +114,9 @@ def _make_fs(
             remote_config = None
 
         conf = self.def_repo.get("config")
-        if conf:
+        if isinstance(conf, dict):
+            config = conf  # type: ignore[unreachable]
+        elif conf:
```
Comment (skshetry) [inline dvc/dependency/repo.py:119]: Can you rewrite this as:

```python
conf  = self.def_repo.get("config", {})
if isinstance(conf, dict):
    config = conf
else:
    config = Config.load_file(conf)
```

Comment (skshetry) [reply dvc/dependency/repo.py:119]: i.e. removing the last branch.

File: dvc/dependency/repo.py [hunk 3]
```diff
@@ -114,7 +114,9 @@ def _make_fs(
             remote_config = None
 
         conf = self.def_repo.get("config")
-        if conf:
+        if isinstance(conf, dict):
+            config = conf  # type: ignore[unreachable]
```
Comment (skshetry) [inline dvc/dependency/repo.py:118]: Also need to update the type for `def_repo` from `Dict[str, str]` instead of ignoring here.

Comment (skshetry) [reply dvc/dependency/repo.py:118]: You can either use a `TypedDict` or just type it as `Dict[str, Any]` (and lose type of values).

**Comment by sjawhar** [comment]
I'll give this branch a test today, thanks for the awesomely quick turnaround! I actually tried doing something very similar to the proposed change in this PR, but I don't think it worked because of either overwriting or otherwise not properly merging with the upstream repo's config.

**Comment by sjawhar** [comment]
So I gave it a test as promised. It's still not working as I was expecting, but perhaps I'm using it wrong. Specifically, I'm trying to override the default remote in the upstream repo _AND_ add config for that remote. The problem is that the import gets created with the following config:
```yaml
    config:
      remote:
        data_registry_nas:
          user: sjawhar
      cache:
        protected: false
        slow_link_warning: true
        verify: false
        dir: /app/.dvc/cache
```

It's missing the `core.remote` setting, so the default remote doesn't get overridden. If I run with `--no-exec` and then add that section to the import, it works. This workaround didn't work before this PR, though, so at least that is fixed! :pray:

It also works if I use `--config` instead of `--remote ... --remote-config`.   But given that `--remote` changes the default remote and `--remote-config` adds configs for the default remote, it seems strange to me that using them together does NOT change the default remote.

**Comment by efiop** [comment]
@sjawhar You are absolutely right. Created https://github.com/iterative/dvc/pull/9855 , please give it a try.

**Pull request merged**